src/main/scala/akka/stream/alpakka/xlsx/WorkbookStreamer.scala

```diff
@@ -1,30 +1,42 @@
 package akka.stream.alpakka.xlsx
 
-import java.io.InputStream
 import java.util.zip.ZipFile
 
 import akka.stream.Materializer
-import akka.stream.alpakka.xml.{ EndElement, ParseEvent, StartElement }
 import akka.stream.alpakka.xml.javadsl.XmlParsing
-import akka.stream.scaladsl.StreamConverters
+import akka.stream.alpakka.xml.{EndElement, ParseEvent, StartElement}
+import akka.stream.contrib.ZipInputStreamSource.ZipEntryData
+import akka.stream.scaladsl.{Source, StreamConverters}
+import akka.util.ByteString
 
 import scala.concurrent.Future
 import scala.util.Try
 
 object WorkbookStreamer {
 
-  def readWorkbook(file: ZipFile)(implicit materializer: Materializer): Future[Map[String, Int]] = {
-    Option(file.getEntry("xl/workbook.xml")) match {
-      case Some(entry) => read(file.getInputStream(entry))
-      case None        => Future.failed(new Exception("invalid xlsx file"))
+  private final val EntryName = "xl/workbook.xml"
+  private final val InvalidFileExceptionMsg = "Invalid xlsx file - no workbook entry"
+
+
+  def readWorkbook(zipFile: ZipFile)(implicit materializer: Materializer): Future[Map[String, Int]] = {
+    Option(zipFile.getEntry(EntryName)) match {
+      case Some(entry) => read(StreamConverters.fromInputStream(() => zipFile.getInputStream(entry)))
+      case None        => Future.failed(new Exception(InvalidFileExceptionMsg))
```

Comment:
You can think of a more specific exception type here. I'd say it should be a runtime one at the very least.

Reply:
In fact, this method fails if the `xl/workbook.xml` file is not present in the *.xlsx (thus *.zip) file, so the `FileNotFoundException` would be nice IMHO (although it's not a runtime one).

Reply:
Does not matter that much in Scala :smile: